src/main/java/software/amazon/smithy/lsp/SmithyLanguageServer.java

```diff
@@ -417,46 +413,72 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
         // or the smithy-build.json itself was changed
-        Set<String> createdSmithyFiles = new HashSet<>(params.getChanges().size());
-        Set<String> deletedSmithyFiles = new HashSet<>(params.getChanges().size());
-        boolean changedBuildFiles = false;
-        for (FileEvent event : params.getChanges()) {
-            String changedUri = event.getUri();
-            if (changedUri.endsWith(".smithy")) {
-                if (event.getType().equals(FileChangeType.Created)) {
-                    createdSmithyFiles.add(changedUri);
-                } else if (event.getType().equals(FileChangeType.Deleted)) {
-                    deletedSmithyFiles.add(changedUri);
-                }
-            } else if (changedUri.endsWith(ProjectConfigLoader.SMITHY_BUILD)
-                    || changedUri.endsWith(ProjectConfigLoader.SMITHY_PROJECT)) {
-                changedBuildFiles = true;
-            } else {
-                for (String extFile : ProjectConfigLoader.SMITHY_BUILD_EXTS) {
-                    if (changedUri.endsWith(extFile)) {
-                        changedBuildFiles = true;
-                        break;
-                    }
-                }
+
+        Map<String, ProjectChanges> changesByProject = projects.computeProjectChanges(params.getChanges());
+
+        changesByProject.forEach((projectName, projectChanges) -> {
+            Project project = projects.getProjectByName(projectName);
+            if (projectChanges.hasChangedBuildFiles()) {
+                client.info("Build files changed, reloading project");
+                // TODO: Handle more granular updates to build files.
+                tryInitProject(projectName, project.root());
+            } else if (projectChanges.hasChangedSmithyFiles()) {
+                Set<String> createdUris = projectChanges.createdSmithyFileUris();
+                Set<String> deletedUris = projectChanges.deletedSmithyFileUris();
+                client.info("Project files changed, adding files "
+                            + createdUris + " and removing files " + deletedUris);
+
+                // We get this notification for watched files, which only includes project files,
+                // so we don't need to resolve detached projects.
+                project.updateFiles(createdUris, deletedUris);
             }
+        });
+
+        // TODO: Update watchers based on specific changes
+        unregisterSmithyFileWatchers().thenRun(this::registerSmithyFileWatchers);
+
+        sendFileDiagnosticsForManagedDocuments();
+    }
+
+    @Override
+    public void didChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams params) {
+        LOGGER.info("DidChangeWorkspaceFolders");
+
+        Either<String, Integer> progressToken = Either.forLeft(UUID.randomUUID().toString());
+        try {
+            client.createProgress(new WorkDoneProgressCreateParams(progressToken)).get();
+        } catch (ExecutionException | InterruptedException e) {
```

Comment:
Why do we block on creating this progress token?

Reply:
It is async in that it sends a request to the client, but I'm not sure if that was your question. I want this method to be mostly synchronous, since we're mutating the project states

Reply:
You indirectly answered it. I was mostly wondering if we were OK with the blocking, and it seems like we are.